### PR TITLE
Add emacs config file and references to README.md

### PR DESCRIPTION
**Pull request by andreaslindner**
What do you guys think about these additions?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -193,9 +144,76 @@ Key:
...
+for feature branches):
+  1. Fork
+  2. Do a good mess
+  3. Merge in feature branch after history rewrite
+
+
+
+## References
+
+* A. Lindner, R. Guanciale och R. Metere, **"TrABin : Trustworthy analyses of binaries"**, Science of Computer Programming, vol. 174, p. 72-89, 2019. [Link](https://doi.org/10.1016/j.scico.2019.01.001). (the proof-producing binary analysis framework with weakest preconditions in HOL4)
```
Comment (totorigolo) [inline README.md]: "och"?

File: scripts/setup/config.emacs [hunk 2]
```diff
@@ -0,0 +1,13 @@
+(add-to-list 'load-path (substitute-in-file-name "${HOLBA_OPT_DIR}/hol_k12_holba/tools"))
+
+(cua-mode)
+
+(load "hol-mode.el")
+;(load "hol-unicode.el")
```
Comment (totorigolo) [inline scripts/setup/config.emacs]: If this mode/script/whatever automatically translates ASCII operators in the equivalent Unicode ones, you could add a comment saying that this is against HolBA coding style.

File: README.md (hunk 1 above)
Comment (andreaslindner) [reply README.md]: ooops, i copied from a KTH page and didn't see this. i fixed some other places but this i missed

File: scripts/setup/config.emacs (hunk 2 above)
Comment (andreaslindner) [reply scripts/setup/config.emacs]: That's a good point. I'll check later.

**Comment by andreaslindner** [comment]
> I think the additions are great. With regard to the README.md in general, I think we should split it up into a README.md and a CONTRIBUTING.md with contribution guidelines. But this is hardly an urgent matter.

This is also a very good point, I'll do this when I fix the Swedish reference texts to English and double-check the emacs file. Do you use a different emacs config?

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  - `snake_case` (e.g. `bir_number_of_mem_splits_def`)
+
+
+
+## Branch policy
+
+### tags
+
+tags should have as many **completed features** as possible:
+ - no cheat
```
Comment (totorigolo) [inline CONTRIBUTING.md]: **See next comment about cheat in `master` before.**

Is that possible? If you tag on master, where development happens, then maybe there is some `cheat`?

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+ - must correctly compile
+ - self tests must succeed
```
Comment (totorigolo) [inline CONTRIBUTING.md]:
```suggestion
 - all tests must succeed, which is checked by Travis CI
```

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ - code should be tested
+
+Follow these instructions whenever you merge to master:
+  - `grep` for "cheat"
```
Comment (totorigolo) [inline CONTRIBUTING.md]: Is that still true, now that master is our dev branch, and that we should merge often?

File: CONTRIBUTING.md [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+  - check that the `README` is up to date (especially tool status)
+  - find a reviewer for your Pull Request
+
+### `master` branch
```
Comment (totorigolo) [inline CONTRIBUTING.md]: This title should be before `Follow these instructions whenever you merge to master:`

File: CONTRIBUTING.md [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+
+`master` is the branch where every feature is available, but not necessarily finalized:
+  - Can cheat, but has to be avoided
```
Comment (totorigolo) [inline CONTRIBUTING.md]: Contradicts what is said above.

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+  - Code should not be be commented out
+  - **Holmake must work**, which is checked by our Travis CI
```
Comment (totorigolo) [inline CONTRIBUTING.md]: Should we add that all tests must pass? That eventually shouldn't be the case, but we cannot really define failing tests for the time being. An alternative would be to always manually check the failing tests in Travis.

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  - bug-fixes are ok
+  - 1 review is needed in order to merge into `master`
```
Comment (totorigolo) [inline CONTRIBUTING.md]:
```suggestion
  - at least 1 review is needed in order to merge into `master`
```

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  - check that the `README` is up to date (especially tool status)
+  - find a reviewer for your Pull Request
+
+### `master` branch
+
+`master` is the branch where every feature is available, but not necessarily finalized:
+  - Can cheat, but has to be avoided
+  - Code should not be be commented out
+  - **Holmake must work**, which is checked by our Travis CI
+  - bug-fixes are ok
```
Comment (totorigolo) [inline CONTRIBUTING.md]:
```suggestion
  - bug-fixing commits are ok
```
Is that what you mean?

File: CONTRIBUTING.md [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+  - 1 review is needed in order to merge into `master`
+
+However, **no development happens on this branch**, but rather on separate
```
Comment (totorigolo) [inline CONTRIBUTING.md]:
```suggestion
However, **no development happens on the `master` branch**, but rather on separate
```

File: README.md [hunk 12]
```diff
@@ -193,9 +144,15 @@ Key:
...
-## Coding style
 
-* HOL source code
-  - Spaces, no tabs
-  - No unicode
-  - `snake_case` (e.g. `bir_number_of_mem_splits_def`)
+
+## References
+
+* A. Lindner, R. Guanciale and R. Metere, **"TrABin : Trustworthy analyses of binaries"**, Science of Computer Programming, vol. 174, p. 72-89, 2019. [Link](https://doi.org/10.1016/j.scico.2019.01.001). (the proof-producing binary analysis framework with weakest preconditions in HOL4)
```
Comment (totorigolo) [inline README.md]:
```suggestion
* A. Lindner, R. Guanciale and R. Metere, **"TrABin : Trustworthy analyses of binaries"**, Science of Computer Programming, vol. 174, p. 72-89, 2019. [Link](https://doi.org/10.1016/j.scico.2019.01.001). _(the proof-producing binary analysis framework with weakest preconditions in HOL4)_
```

File: README.md [hunk 13]
```diff
@@ -193,9 +144,15 @@ Key:
...
+
+* D. Lundberg, **"Provably Sound and Secure Automatic Proving and Generation of Verification Conditions"**, Master Thesis, 2018. [Link](http://urn.kb.se/resolve?urn=urn%3Anbn%3Ase%3Akth%3Adiva-239441).
+
+* R. Metere, A. Lindner and R. Guanciale, **"Sound Transpilation from Binary to Machine-Independent Code"**, in 20th Brazilian Symposium on Formal Methods, p. 197-214, 2017. [Link](https://link.springer.com/chapter/10.1007/978-3-319-70848-5_13). (formalization of intermediate language and proof-producing lifter in HOL4)
```
Comment (totorigolo) [inline README.md]:
```suggestion
* R. Metere, A. Lindner and R. Guanciale, **"Sound Transpilation from Binary to Machine-Independent Code"**, in 20th Brazilian Symposium on Formal Methods, p. 197-214, 2017. [Link](https://link.springer.com/chapter/10.1007/978-3-319-70848-5_13). _(formalization of intermediate language and proof-producing lifter in HOL4)_
```

File: CONTRIBUTING.md (hunk 5 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: let's fix a test that checks exactly this, and i would say yes

Comment (totorigolo) [reply CONTRIBUTING.md]: How do you want to check that? Is there a wait to forbid cheating with Holmake?

File: CONTRIBUTING.md (hunk 3 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: let's start with the good intentions and then we can gradually go back in case it is hard to keep them up. in case it is hard to follow up on this, we could tag instable modules or directories, where tests are not required to succeed (for example)

File: CONTRIBUTING.md (hunk 6 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: yes

File: CONTRIBUTING.md (hunk 3 above)
Comment (totorigolo) [reply CONTRIBUTING.md]: :+1: That's what I had in mind: tag and mention tools status.

File: CONTRIBUTING.md (hunk 7 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: that is true. i think we wrote this here together and didn't fix the other one. and this is also the reason why you asked for the other one now. until now we didn't need cheat in there, so let's see how long we can keep it like this.

File: CONTRIBUTING.md (hunk 8 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: good point, let's just keep it in mind for now and see how it goes

File: CONTRIBUTING.md (hunk 5 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: no, but it could grep cheat 😄

File: CONTRIBUTING.md (hunk 3 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: let's build the support for this later and just mention it here for now. so we settle for a tagging of unstable directories which enables usage of cheat and failing tests in there

Comment (andreaslindner) [reply CONTRIBUTING.md]: and we build it as soon as we need it

File: CONTRIBUTING.md (hunk 5 above)
Comment (totorigolo) [reply CONTRIBUTING.md]: Yeah but that doesn't work, just try it on your local repository. However, you can grep the build logs, Holmake writes "CHEATED" when that's the case.

Comment (andreaslindner) [reply CONTRIBUTING.md]: then this means grepping sml files only

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+   - All tests must succeed
+ - Bug-fixing commits are ok
+ - At least 1 accepting review is needed in order to merge into `master`
+
+Notice:
+ - **No development happens on the `master` branch**, but rather on separate feature branches
+ - **In order to prevent mayhem**, define good interfaces for your code (so that development won't break existing code)
+
+Follow these instructions whenever you merge to master:
+  - `grep` for "cheat" (TODO: add this as a CI test)
```
Comment (totorigolo) [inline CONTRIBUTING.md]: Prefer a new issue instead of this TODO that we will forget.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+  - `snake_case` (e.g. `bir_number_of_mem_splits_def`)
+
+
+
+# Branch policy
+
+## `master` branch
+
+`master` is the branch where every feature is available, but not necessarily finalized:
+ - Can cheat, but has to be avoided (allowed only in a small number of modules, which are explicitly marked unstable, no-cheat CI test mentioned below would need a corresponding exception list)
```
Comment (totorigolo) [inline CONTRIBUTING.md]: I didn't find anything about **no-cheat** CI tests specifically. Did you forget it?

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+## Feature branches
+
+Every "somewhat" working tool should be available in the `master` branch, but new
+features or any development must go on new branches prefixed with `dev_`, fixes with `fix_`.
+
+Guidelines:
+ - Branch names must be short and explicit (prefer explicit over short)
+ - Every feature branch should involve small developments
+ - Rebase feature branches on `master` **often**, by using `git rebase` or `git merge`
+ - **Merge feature branches to `master` often**: work on small features
```
Comment (totorigolo) [inline CONTRIBUTING.md]: So, should we merge or rebase often?

Comment (andreaslindner) [reply CONTRIBUTING.md]: I wonder how these two similar bullet points got there. It was just moved from the README.md file originally.

File: CONTRIBUTING.md (hunk 15 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: no, but it wasn't as wordy as you expected it possibly :)

File: CONTRIBUTING.md (hunk 14 above)
Comment (andreaslindner) [reply CONTRIBUTING.md]: done, and kept as TODOs. pushing now
